lib/interface-types/src/decoders/binary.rs

```diff
@@ -0,0 +1,940 @@
+            output_types: import_output_types,
+        });
+    }
+
+    Ok((input, imports))
+}
+
+/// Parse a list of adapters.
+fn adapters<'input, E: ParseError<&'input [u8]>>(
+    input: &'input [u8],
+) -> IResult<&'input [u8], Vec<Adapter>, E> {
+    let mut input = input;
+
+    consume!((input, number_of_adapters) = leb(input)?);
+
+    let mut adapters = Vec::with_capacity(number_of_adapters as usize);
+
+    for _ in 0..number_of_adapters {
+        consume!((input, adapter_kind) = byte(input)?);
+        let adapter_kind = AdapterKind::try_from(adapter_kind)
+            .map_err(|_| Err::Error(make_error(input, ErrorKind::ParseTo)))?;
+
+        match adapter_kind {
+            AdapterKind::Import => {
+                consume!((input, adapter_namespace) = string(input)?);
+                consume!((input, adapter_name) = string(input)?);
+                consume!((input, adapter_input_types) = list(input, ty)?);
+                consume!((input, adapter_output_types) = list(input, ty)?);
+                consume!((input, adapter_instructions) = list(input, instruction)?);
+
+                adapters.push(Adapter::Import {
+                    namespace: adapter_namespace,
+                    name: adapter_name,
+                    input_types: adapter_input_types,
+                    output_types: adapter_output_types,
+                    instructions: adapter_instructions,
+                });
+            }
+
+            AdapterKind::Export => {
+                consume!((input, adapter_name) = string(input)?);
+                consume!((input, adapter_input_types) = list(input, ty)?);
+                consume!((input, adapter_output_types) = list(input, ty)?);
+                consume!((input, adapter_instructions) = list(input, instruction)?);
+
+                adapters.push(Adapter::Export {
+                    name: adapter_name,
+                    input_types: adapter_input_types,
+                    output_types: adapter_output_types,
+                    instructions: adapter_instructions,
+                });
+            }
+
+            AdapterKind::HelperFunction => {
+                consume!((input, adapter_name) = string(input)?);
+                consume!((input, adapter_input_types) = list(input, ty)?);
+                consume!((input, adapter_output_types) = list(input, ty)?);
+                consume!((input, adapter_instructions) = list(input, instruction)?);
+
+                adapters.push(Adapter::HelperFunction {
+                    name: adapter_name,
+                    input_types: adapter_input_types,
+                    output_types: adapter_output_types,
+                    instructions: adapter_instructions,
+                });
+            }
+        }
+    }
+
+    Ok((input, adapters))
+}
+
+/// Parse a list of forwarded exports.
+fn forwards<'input, E: ParseError<&'input [u8]>>(
+    input: &'input [u8],
+) -> IResult<&'input [u8], Vec<Forward>, E> {
+    let mut input = input;
```

Comment:
⛳️ I believe you can just put the `mut` in the argument list, like `mut input: &'input [u8],`

Reply:
True! Addressed in https://github.com/wasmerio/wasmer/pull/787/commits/d8f923091f89686faa99fc7a8af378873f86c514.